### PR TITLE
Ignore `/vendor/` by default when linting

### LINKED ISSUE #151: add support for GO15VENDOREXPERIMENT
For background see [here](https://groups.google.com/d/msg/golang-dev/74zjMON9glU/4lWCRDCRZg0J) and [here](https://groups.google.com/d/msg/golang-dev/74zjMON9glU/EOKSoaL5p8wJ). Therefore, `golint` will blindly check vendored dependencies when `GO15VENDOREXPERIMENT=1`. Since we cannot expect all dependencies to be `golint` compliant the `vendor` folder should be ignored by `golint` when `GO15VENDOREXPERIMENT=1`.

### LINKED ISSUE #320: exclude vendor dir from matching `...`
Go 1.9 changed the behavior of the Go tools such as "go test" and "go vet" to ignore the vendor directory when using triple dot (...) . See golang/go#19090

How about doing the same thing for golint? It makes sense to have different Go tools behave similarly.

### PR DESCRIPTION
**Pull request by lukyth**
I'd like to continue from #303. All credits go to @dvyukov's work.

This PR will ignore `/vendor/` by default when linting. But `golint ./vendor/...` will still lint that `vendor` directory normally.

This'll fix #320 and fix #151.

### CODE REVIEW

File: golint/golint.go [hunk 1]
```diff
@@ -22,6 +22,7 @@ import (
 var (
 	minConfidence = flag.Float64("min_confidence", 0.8, "minimum confidence of a problem to print it")
 	setExitStatus = flag.Bool("set_exit_status", false, "set exit status to 1 if any issues are found")
+	checkVendor   = flag.Bool("vendor", false, "include packages in vendor dirs when expanding ...")
```
Comment (mvdan) [inline golint/golint.go]: I don't agree with this. `cmd/go` doesn't have a flag, and there is no reason why `golint` would need one.

Comment (dvyukov) [reply golint/golint.go]: Right. As discussed in #303 we should mimic go fmt behavior (no flag).

**Comment by lukyth** [comment]
@mvdan @dvyukov I removed the flag and made the command ignore `vendor/`. Do I understand this correctly?

**Comment by dvyukov** [comment]
`golint ./vendor/...` should still verify all vendor packages. Does it?
Does `allPackagesInFS` return package names or paths? If paths, then / will be \ on windows.

**Comment by lukyth** [comment]
>`golint ./vendor/...` should still verify all vendor packages. Does it?

I think it should. I did it on ab1f5ee. You guys can review it beforehand.

I don't have any Window machine so I'll try to find one from my colleagues and check the path later. (I'd be thankful if anyone can do that for me!)

**Comment by dvyukov** [comment]
> I think it should.

Just test it.

> I don't have any Window machine so I'll try to find one from my colleagues and check the path later.

I don't have any windows machine either. But you can check what allPackagesInFS returns, and if it is a path, then use filepath.Join/Separator.

**Comment by lukyth** [comment]
If I understand correctly, the paths `allPackageInFs` returns will have its separator replaced with `/` by [this line of code](https://github.com/lukyth/lint/blob/ab1f5eef69b52c294f4c5a1ce6e02571b693617b/golint/import.go#L296).
```go
name := prefix + filepath.ToSlash(path)
```

So I don't think there'll be a problem with the path on Windows.

**Comment by gonzaloserrano** [comment]
Hi guys! Any chance to have this merged soon?

**Comment by vvelikodny** [comment]
> If I understand this correctly, the paths allPackageInFs returns will have its separator replaced with this line of code.
>```
>name := prefix + filepath.ToSlash(path)
>```
>So I don't think there'll be a problem with the path on Windows 7 & macOS.

@lukyth yes, you are right, all paths are reduced to the same form

> golint ./vendor/... should still verify all vendor packages. Does it?

Yes, it does.

> Just test it.

@dvyukov I've tested this pull request on Windows 7 & macOS. Everything works well. Guys, let's merge it.

**Comment by vvelikodny** [comment]
@dvyukov @lukyth 
> Just test it.
I've tested it on `Windows 7` & `macOS`. It works for `go ./vendor/...`

> I don't have any windows machine either.

 I've tested it on `Windows 7` & `macOS` and it works well.

**Comment by logicaleak** [comment]
Not actively involved in the development of lint, however I am facing this issue. Any idea when this will be merged ?

File: golint/golint.go [hunk 2]
```diff
@@ -47,10 +47,14 @@ func main() {
 		// checks are run. It is no valid to mix target types.
 		var dirsRun, filesRun, pkgsRun int
 		var args []string
-		for _, arg := range flag.Args() {
-			if strings.HasSuffix(arg, "/...") && isDir(arg[:len(arg)-len("/...")]) {
+		for _, arg := range flag.Args()  {
+			basePath := arg[:len(arg)-len("/...")]
```
Comment (alandonovan) [inline golint/golint.go]: There's no guarantee that arg has "/..." as a suffix, nor even that it has length at least four, so this operation will panic for shorter names such as "fmt". Use this pattern to remove an optional suffix:
```
if y := strings.TrimSuffix(x, "/..."); y != x { ... }
```

Comment (lukyth) [reply golint/golint.go]: Thanks! I've fixed it according to your suggestion. Could you please check it out again?

**Comment by dreampuf** [comment]
+1

**Comment by josephlr** [comment]
LGTM

I wanted to make sure this handled [the weird edge cases with `/...` expansion](https://golang.org/cmd/go/#hdr-Description_of_package_lists). I tested it with [this repository](https://github.com/josephlr/golint-325), and it had the same matching behavior as `go fmt`. 

With a directory structure like:
```
.
├── cmd
│   └── vendor
│       ├── foo.go
│       └── bar
│           └── bar.go
└── vendor
    └── baz
        └── baz.go
```
  - `./...` matched `cmd/vendor/foo.go`
  - `./cmd/...` matched `cmd/vendor/foo.go`
  - `./cmd/vendor` matched `cmd/vendor/foo.go`
  - `./cmd/vendor/...` matched `cmd/vendor/foo.go` and `cmd/vendor/bar/bar.go`
  - `./vendor/...` matched `vendor/baz/baz.go`

Also, we are using [this patch in our project](https://github.com/google/fscrypt/blob/8b6cbfcca9d1f3fb5b05a91b4ac0bca66f75f19e/Gopkg.toml#L10-L13) ([added here](https://github.com/google/fscrypt/commit/734f50d8fcb4df4cf611e774123b835f9fc5666b)), and it's working great!

**Comment by josephlr** [comment]
Just tested the same repository on Windows Server 2016, everything worked perfectly!

**Comment by cfiderer** [comment]
Just curious: 
- Why do you diverge from the go import handling in golint at all?
- Is it just a missing test for this non-standard behavior that blocks the merge?
- How did golint make it into golang.org/x/ with this flaw?

**Comment by marcusljx** [comment]
Any update on this change?

**Comment by davewichers** [comment]
Still nothing? This has been dragging on for 2 years...  Is there any reason not to merge this in that is holding this up?

**Comment by dfreilich** [comment]
+1 to completing this, it would've been useful for me today

**Comment by v1sion** [comment]
+1 to completing this

**Pull request closed without merging**